ARCs/arc-0060.md

```diff
@@ -152,298 +182,206 @@ export interface HDWalletMetadata {
+  return toCanonicalJson(data);
+}
+```
 
-- the `signingData` and `metadata` **MUST NOT** be `null`, otherwise the wallet **MUST** reject the call.
-- `signingData.data` **MUST** be valid respect the JSON schema in `metadata.schema`. If the validation fails, the call **MUST** be rejected with a `4300` error.
-- if the encoding `metadata.encoding` is present, it **MUST** be used to decode `signingData.data.bytes`.
-- the wallet **MUST** prepare the signing message as `msg=(<signingData.data.bytes>)`.
-  - if the value of `ARC60Domain` is not empty than it **MUST** be used as `prefix`: `msg=(<prefix> || <signingData.data.bytes>)`.
-- the wallet **MUST** ask users for signing confirmation. It **MUST** display a message (defined in Display message) providing the `ScopeType`, `signer`, and the `msg` being signed:
-  - if the user approves, `msg` **MUST** be signed with the `signer`'s key and `ret` **MUST** be set to the corresponding `Signature`.
-  - if the user rejects, the call **MUST** fail with error code `4001`.
 
-Note that if `msg` cannot be signed for any reason, the wallet **MUST** throw an error, such that
+## Semantic Requirements
 
-- `err.message` **SHOULD** indicate the reason for the error (e.g. specify that `signingData.data` is not a valid JSON object according to `metadata.schema`)
-- `err.failingSignData` **SHOULD** return the `StdSigData` object that caused the error, otherwise `null`.
+Upon calling `signData(signingData)`:
+- the wallet **MUST** parse, validate, and serialize `signingData.data` as defined by the schema for scope `signingData.scope`
+- the wallet **MUST** ask users for signing confirmation. It **MUST** display a message (defined in section Display message below) providing the `scope`, `signer`, and the `data` being signed:
+  - if the user approves, `data` **MUST** be signed with the `signer`'s key and the resulting `Signature` must be returned.
+  - if the user rejects, the call **MUST** fail with error code `4001`.
 
-#### Semantic of `StdSigData`
+### Semantics of `StdSignData`
 
-- it **MUST** be a valid `StdSigData` object, otherwise the wallet **MUST** reject.
+- it **MUST** be a valid `StdSignData` object, otherwise the wallet **MUST** reject.
 - `data`:
-  - it **MUST** be a valid `StdDataStr` object (UTF-8 encoded), otherwise the wallet **MUST** reject.
-  - it **MUST** adhere to `metadata.schema`, otherwise the wallet **MUST** reject.
-  - the `data.ARC60Domain` field must be validated against forbidden values:
-    - the wallet **MUST** reject if `ARC60Domain="TX"`.
-    - the wallet **MUST** reject if `ARC60Domain="TG"`.
-  - `data.bytes` must be a valid byte string:
-    - it **MUST** be encoded as `metadata.encoding` if present, otherwise `base64`.
-    - the wallet **MUST** throw a `4602` error if the decoding fails with the given encoding value.
-    - the wallet **MUST** reject if `data.bytes` field is prepended with a forbidden domain separator `TX` or `TG`.
```

Comment:
In this version of the ARC we never mention forbidden domain separators like `TX` or `TG`. We should do that unless we want people to start using ARC60 to sign valid transactions.

Reply:
I think it's the schema's responsibility to make sure that the object being signed doesn't serialize to something that's a transaction, logicsig, etc.

For example, the `CHALLENGE32` scope that we talked about has 32-byte payloads and is too short to encode a valid transaction, hence it'd be okay if a payload starts with `TX` or `TG`. If we enforce that a payload cannot have `TX` or `TG` as prefix, we cannot support truly random payload (like a random nonce). It was my impression that this needs to be supported.